src/coreclr/jit/valuenum.cpp

```diff
@@ -4594,6 +4594,97 @@ bool ValueNumStore::IsVNConstant(ValueNum vn)
     }
 }
 
+bool ValueNumStore::IsVNVectorZero(ValueNum vn)
+{
+    if (vn == NoVN)
+    {
+        return false;
+    }
+#if FEATURE_HW_INTRINSICS
+    // REVIEW: This will only return true if Vector.Zero encodes
+    //         its base type as an argument. On XARCH there may be
+    //         scenarios where Vector.Zero will not encode its base type;
+    //         therefore, returning false here.
+    VNFuncApp funcApp;
+    if (GetVNFunc(vn, &funcApp) && funcApp.m_arity == 1)
+    {
+        switch (funcApp.m_func)
+        {
+            case VNF_HWI_Vector128_get_Zero:
+#if defined(TARGET_XARCH)
+            case VNF_HWI_Vector256_get_Zero:
+#elif defined(TARGET_ARM64)
+            case VNF_HWI_Vector64_get_Zero:
+#endif
+            {
+                return true;
+            }
+
+            default:
+                return false;
+        }
+    }
+#endif
+    return false;
+}
+
+#if FEATURE_SIMD
+VNSimdTypeInfo ValueNumStore::GetSimdTypeOfVN(ValueNum vn)
+{
+    VNSimdTypeInfo vnInfo;
+
+    if (vn == NoVN)
+    {
+        vnInfo.m_simdSize        = 0;
+        vnInfo.m_simdBaseJitType = CORINFO_TYPE_UNDEF;
+        return vnInfo;
+    }
+
+    // The SIMD type is encoded as a function,
+    // even though it is not actually a function.
+    VNFuncApp simdType;
+    if (GetVNFunc(vn, &simdType) && simdType.m_func == VNF_SimdType)
+    {
+        assert(simdType.m_arity == 2);
+        vnInfo.m_simdSize        = GetConstantInt32(simdType.m_args[0]);
+        vnInfo.m_simdBaseJitType = (CorInfoType)GetConstantInt32(simdType.m_args[1]);
+        return vnInfo;
+    }
+
+    vnInfo.m_simdSize        = 0;
+    vnInfo.m_simdBaseJitType = CORINFO_TYPE_UNDEF;
+    return vnInfo;
+}
+
+// REVIEW: Vector.Zero nodes in VN currently encode their SIMD type for
+//         conservative reasons. In the future, it might be possible not do this
+//         on most platforms since Vector.Zero's base type does not matter.
+VNSimdTypeInfo ValueNumStore::GetVectorZeroSimdTypeOfVN(ValueNum vn)
+{
+    VNSimdTypeInfo vnInfo;
+
+    if (!IsVNVectorZero(vn))
+    {
+        vnInfo.m_simdSize        = 0;
+        vnInfo.m_simdBaseJitType = CORINFO_TYPE_UNDEF;
+        return vnInfo;
+    }
+
+    // Vector.Zero does not have any arguments,
+    // but its SIMD type is encoded as an argument.
+    VNFuncApp funcApp;
+    if (GetVNFunc(vn, &funcApp))
+    {
+        assert(funcApp.m_arity == 1);
+        return GetSimdTypeOfVN(funcApp.m_args[0]);
+    }
```

Comment:
I think it would be nice to unify `IsVNVectorZero` and this function into one that has an output parameter since now there is a dependency between them there might not be totally clear.

Reply:
I think we could unify them by effectively making the implementation of `IsVNVectorZero` try to get the simd type.